slam/src/slam/fast_slam2/FastSLAM2Particle.py

```diff
@@ -0,0 +1,246 @@
+    Correct particle pose using measurements
+    Compute particle weight
+    '''
+    self._log('Update Measurement')
+    self._log('before measurement - particle pose',self.pose)
+    # ++++++ Set up: sort meas_ls by range; initialize particle mean, covariance, and weight
+    meas_ls.sort(key = lambda meas: meas.data[0])  # meas_ls should be sorted by range.
+    pose_mean, pose_cov = np.copy(self.pose), np.copy(self.pose_cov)
+    self.weight = 1.0
+
+    # ++++++ No Measurment: sample the pose using motion model
+    if len(meas_ls) == 0:
+      self.pose = self._robot_physics.compute_sample_pose(pose_mean, pose_cov)
+      self._log('no measurement - particle pose',self.pose)
+      return
+
+    # ++++++ Access each measurement 
+    observed_landmark = None
+    
+    # Iteratively update pose mean and pose covariance using each measurement of the same time step
+    for meas in meas_ls:
+      # ++++++ If unknown correspondence, perform data association for each landmark 
+      if meas.correspondence is None:
+        for land_idx, landmark in list(self.landmarks.items()):
+          # +++ New particle pose is computed here
+          self._update_dependencies_on_landmark(landmark, pose_mean, pose_cov, meas)
+          self._update_pose_distribution(landmark, pose_mean, pose_cov)
+          self._update_landmark_association_prob(landmark)
+          
+        if len(self.landmarks) > 0: # use maximum likelihood to select landmark
+          observed_landmark = max(list(self.landmarks.values()), key = lambda landmark: landmark.association_prob)
+      
+        # No landmark or not confident: create new landmark 
+        if observed_landmark is None or ( \
+              observed_landmark.association_prob < self._landmark_constants.new_landmark_threshold and \
+              not self._are_landmarks_fixed \
+          ):
+          key = len(self.landmarks) + 1
+          observed_landmark = self.landmarks[key] = \
+            self._init_landmark_with_meas(pose_mean, pose_cov, meas, name=str(key))
+
+        else:
+          self._update_observed_landmark(observed_landmark, pose_cov, meas)  
+
+      # ++++++ If known correspondence
+      else:
+        if meas.correspondence in self.landmarks:
+          # +++ New particle pose is computed here
+          observed_landmark = self.landmarks[meas.correspondence]
+          self._update_dependencies_on_landmark(observed_landmark, pose_mean, pose_cov, meas)
+          self._update_pose_distribution(observed_landmark, pose_mean, pose_cov)
+          self._update_observed_landmark(observed_landmark, pose_cov, meas)
+        else:
+          observed_landmark = self.landmarks[meas.correspondence] = \
+            self._init_landmark_with_meas(pose_mean, pose_cov, meas, name=str(meas.correspondence))
+      # ++++++ Penalize landmarks that we expect to see
+      if not self._are_landmarks_fixed:
+        self._update_unobserved_landmarks(observed_landmark, meas)
```

Comment:
I'm not sure I understand how this logic works. It seems like somehow you need to consider each pair of landmark and measurement, but I'm not totally clear how that's happening. Anyways, some comments explaining the logic of how this finds unobserved landmarks would be helpfu.

Reply:
Agreed. Thanks!